### PR TITLE
improve unittests and fix strptime of status datetime for 3.6

### PR DESCRIPTION
**Pull request by Vitalts**
• mock all requests to real overpass endpoint in unittests
• refactor unittests
• add more unittests
• fix strptime of status datetime for 3.6

### CODE REVIEW

**Comment by mvexel** [comment]
Hi @Vitalts - after merging @dericke 's PR https://github.com/mvexel/overpass-api-python-wrapper/pull/141 there are some minor merge conflicts, could you resolve and then I can merge, please?

**Comment by dericke** [comment]
We've dropped py3.6 support since this PR was opened, so the parts related to that shouldn't be merged.

**Comment by mvexel** [comment]
@dericke @Vitalts how would you like to proceed with this, in light of the significant changes in `test_api.py` in #140 ?

**Comment by dericke** [comment]
I won't be able to spend any time on this for a couple of weeks, but I'll check in and see how I can help out at that point.

File: overpass/api.py [hunk 1]
```diff
@@ -93,7 +95,7 @@ def get(self, query, responseformat="geojson", verbosity="body", build=True, dat
                 date = datetime.fromisoformat(date)
             except ValueError:
                 # The 'Z' in a standard overpass date will throw fromisoformat() off
-                date = datetime.strptime(date, '%Y-%m-%dT%H:%M:%SZ')
+                date = self._strptime(date)
```
Comment (dericke) [inline overpass/api.py:98]: This change is not needed since we've dropped 3.6 support

File: overpass/api.py [hunk 2]
```diff
@@ -137,7 +139,17 @@ def get(self, query, responseformat="geojson", verbosity="body", build=True, dat
         return self._as_geojson(response["elements"])
 
     @staticmethod
-    def _api_status() -> dict:
+    def _strptime(date_string):
```
Comment (dericke) [inline overpass/api.py:142]: Also deprecated by the dropping of 3.6 support

File: overpass/api.py [hunk 3]
```diff
@@ -204,9 +207,19 @@ def search(self, feature_type, regex=False):
         """Search for something."""
         raise NotImplementedError()
 
+    def __deprecation_get(self, *args, **kwargs):
```
Comment (dericke) [inline overpass/api.py:210]: Everything from this point, in this file, is still relevant with the dropping of 3.6.

File: overpass/dependency.py [hunk 4]
```diff
@@ -0,0 +1,9 @@
+import sys
```
Comment (dericke) [inline overpass/dependency.py:1]: Deprecated by the dropping of 3.6

File: tests/__init__.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
+import os
+
+# Set temporary True to update example.json/example.response files with real overpass endpoint data
+UPDATE_EXAMPLES = False
+
+RESOURCE_PATH = os.path.join(os.path.dirname(os.path.abspath(os.path.relpath(__file__))), 'resources')
```
Comment (dericke) [inline tests/__init__.py:6]: Can we use `pathlib` instead of `os.path`?

**Review by dericke — CHANGES_REQUESTED** [review]

File: overpass/api.py [hunk 6]
```diff
@@ -137,7 +139,17 @@ def get(self, query, responseformat="geojson", verbosity="body", build=True, dat
...
+        if dependency.Python.less_3_7():
+            dt = datetime.strptime(date_string, '%Y-%m-%dT%H:%M:%SZ')
+            kwargs = {k: getattr(dt, k) for k in ('year', 'month', 'day', 'hour', 'minute', 'second', 'microsecond')}
+            kwargs['tzinfo'] = timezone.utc
+            return datetime(**kwargs)
+
+        return datetime.strptime(date_string, '%Y-%m-%dT%H:%M:%S%z')
+
+    @classmethod
+    def _api_status(cls) -> dict:
```
Comment (dericke) [inline overpass/api.py:152]: Also deprecated by the dropping of 3.6 support

File: overpass/api.py [hunk 7]
```diff
@@ -204,9 +207,19 @@ def search(self, feature_type, regex=False):
...
+    def __deprecation_search(self, *args, **kwargs):
+        import warnings
+        warnings.warn('Call to deprecated function "Search", use "search" function instead', DeprecationWarning)
+        return self.search(*args, **kwargs)
+
     # deprecation of upper case functions
-    Get = get
-    Search = search
+    Get = __deprecation_get
+    Search = __deprecation_search
```
Comment (dericke) [inline overpass/api.py:222]: Still relevant with the dropping of 3.6.

File: overpass/api.py [hunk 8]
```diff
@@ -154,24 +166,15 @@ def _api_status() -> dict:
...
             if word.startswith('Currently running queries')
         )
         running_slots = tuple(tuple(line.split()) for line in lines[current_idx + 1:])
-        running_slots_datetimes = tuple(
-            datetime.strptime(
-                slot[3], "%Y-%m-%dT%H:%M:%S%z"
-            )
-            for slot in running_slots
-        )
+        running_slots_datetimes = tuple(cls._strptime(slot[3]) for slot in running_slots)
```
Comment (dericke) [inline overpass/api.py:177]: Also deprecated by the dropping of 3.6 support

File: tests/__init__.py [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+
+
+def save_resource(file_name, data):
+    with open(os.path.join(RESOURCE_PATH, file_name), 'wb') as f:
+        f.write(data)
+
+
+def load_resource(file_name):
+    with open(os.path.join(RESOURCE_PATH, file_name), 'rb') as f:
```
Comment (dericke) [inline tests/__init__.py:15]: Can we use pathlib instead of os.path?

**Review by dericke — CHANGES_REQUESTED** [review]

**Comment by dericke** [comment]
Everything in the files I didn't comment on looks to still be relevant at first glance (and great work there!) but I haven't run any of the tests yet to be sure.

File: tests/__init__.py (hunk 9 above)
Comment (mvexel) [reply tests/__init__.py:15]: Not familiar. What does that give us?

**Comment by mvexel** [comment]
@Vitalts are you still interested in getting this over the finish line? I'm sorry it all took so long.. With @dericke 's comments we are one step closer.

File: overpass/api.py (hunk 6 above)
Comment (mvexel) [reply overpass/api.py:152]: you mean we can get rid of the custom `_strptime()` function altogether and replace with plain `datetime.strptime()`?

**Comment by mvexel** [comment]
so @dericke @Vitalts This branch has diverged from main enough that merging will be a bit of a mess. I could use some help especially with `api.py`

Anyone has some time to spend on this? Or should we abandon this PR and start a new one focused on fixing the tests and deprecating 3.6?

**Comment by dericke** [comment]
I would be in favor of closing this PR and opening a new one focused just on the tests.

**Comment by mvexel** [comment]
thanks @dericke , I will close this then and create an issue that we can create a new PR against.

**Pull request closed without merging**